common/src/chain/block/block_body/merkle_proxy.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use merkletree::{
+    proof::single::{SingleProofHashes, SingleProofNodes},
+    tree::MerkleTree,
+};
+
+use crate::primitives::H256;
+
+use super::{
+    block_merkle::{calculate_tx_merkle_tree, calculate_witness_merkle_tree},
+    merkle_tools::MerkleHasher,
+    BlockBody, BlockMerkleTreeError,
+};
+
+#[must_use]
+pub struct BlockBodyMerkleProxy {
+    witness_tree: WrappedWitnessMerkleTree,
+    tree: WrappedMerkleTree,
+}
```

Comment:
It used to take a reference to the body and cache the result on call... but then I fell into mutability issues that couldn't be solved without nasty RefCells and copies, so I decided to just evaluate the trees on construction. I was OK with that given that in all verification cases (so far) we always have to evaluate both.

Reply:
About the reasoning, without such a structure, there will be 4 functions that all sound the same... high probability of errors are there. So I thought I'd create such a structure where it's less likely to confuse things.

Reply:
Sorry I may have been a bit unclear. My concern in the original comment was mainly with the naming rather than the mechanism itself. Not sure where the concept of "proxy" comes in. Not that I have any constructive alternatives...

Reply:
> Not that I have any constructive alternatives...

Neither do I... which is why I kept it like this. It's still a proxy in a remote way, given that a user of the struct doesn't see the internals. It doesn't bother me that much anyway. So, if you have a better name, please go ahead and suggest it :smile:

Reply:
Meh the name's fine for me